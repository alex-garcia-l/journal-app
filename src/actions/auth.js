import Swal from 'sweetalert2';

import { types } from '../types/types';
import { finishLoading, startLoading } from './ui';
import { firebase, googleAuthProvider } from '../database/firebase-config';

export const startLogin = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    firebase.auth().signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch(error => {
        dispatch(finishLoading());
        Swal.fire('Error', error.message, 'error');
      });
  }
}

export const startLoginWithGoogle = () => {
  return (dispatch) => {
    firebase.auth().signInWithPopup(googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
      });
  }
}

export const startUserRegister = (name, email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    firebase.auth().createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch(error => {
        dispatch(finishLoading());
        Swal.fire('Error', error.message, 'error');
      });
  }
}

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout());
  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
})

export const logout = () => ({
  type: types.logout
})
