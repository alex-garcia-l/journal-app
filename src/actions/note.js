import Swal from 'sweetalert2';
import { db } from '../database/firebase-config';
import { types } from '../types/types';
import { loadNotes } from '../helpers/loadNotes';
import { uploadImage } from '../helpers/uploadImage';

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }

    const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
    dispatch(setNoteActive(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote));
  }
}

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  }
}

export const startUpdateNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    if (!note.url) {
      delete note.url;
    }

    const noteUpdate = { ...note };
    delete noteUpdate.id;

    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteUpdate);

    dispatch(refreshNotes(note.id, noteUpdate));

    Swal.fire({
      title: 'Success',
      text: 'Note modified',
      icon: 'success',
      showConfirmButton: false,
      timer: 1500
    });
  }
}

export const startUploadImage = (file) => {
  return async (dispatch, getState) => {
    const { active: note } = getState().notes;

    Swal.fire({
      title: 'Uploading',
      text: 'Please wait...',
      allowOutsideClick: false,
      allowEscapeKey: false,
      allowEnterKey: false,
      didOpen: () => {
        Swal.showLoading();
      },
      willClose: () => {
        Swal.hideLoading()
      },
    });

    const upload = await uploadImage(file);
    note.url = upload.secure_url;
    dispatch(startUpdateNote(note));

    Swal.close();
  }
}

export const startDeleteNote = (id) => {
  return (dispatch, getState) => {

    Swal.fire({
      title: 'Are you sure you want to delete the note?',
      confirmButtonText: 'Yes, delete',
      showCancelButton: true,
      icon: 'question',
    }).then(async (result) => {
      /* Read more about isConfirmed, isDenied below */
      if (result.isConfirmed) {
        const { uid } = getState().auth;

        await db.doc(`${uid}/journal/notes/${id}`).delete();
        dispatch(deleteNote(id));
      }
    })
  }
}

export const setNoteActive = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note
  }
});

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note
  }
});

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes
});

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id
});

export const refreshNotes = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: {
      id,
      ...note
    }
  }
});

export const purgeNotes = () => ({
  type: types.notesLogoutCleaning
})
