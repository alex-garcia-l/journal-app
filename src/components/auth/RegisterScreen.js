import React from 'react';
import { Link } from 'react-router-dom';

export const RegisterScreen = () => {
  return (
    <>
      <h3 className="auth__title mb-5">
        <i class="fa-solid fa-user-plus fa-3x mb-5"></i>
        <span>Register</span>
      </h3>
      
      <form>
        <input
          type="text"
          placeholder="Name"
          name="name"
          autoComplete="off"
          className="input"
        />

        <input
          type="email"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="input"
        />

        <input
          type="password"
          placeholder="Password"
          name="password"
          className="input"
        />

        <input
          type="password"
          placeholder="Confirm password"
          name="password_confirm"
          className="input"
        />

        <button
          type="submit"
          className="btn btn-primary btn-block"
        >
          Register
        </button>

        <div className="auth__create_acount">
          <Link
            to="/auth/login"
          >
            Already registered?
          </Link>
        </div>
      </form>
    </>
  );
};
