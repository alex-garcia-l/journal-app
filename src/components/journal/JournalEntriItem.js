import React from 'react';

export const JournalEntriItem = () => {
  return (
    <div className="journal__entry">
      <div
        className="journal__entry-picture"
        style={{
          backgroundImage: 'url("https://www.publicdomainpictures.net/pictures/320000/nahled/background-image.png")'
        }}
      ></div>

      <div className="journal__entry-container">
        <div className="journal__entry-body">
          <p className="journal__entry-title">
            Un nuevo d√≠a
          </p>

          <p className="journal__entry-content">
            Lorem ipsum dolor sit amet, consectetur adipiscing elit.
          </p>
        </div>

        <div className="journal__entry-date-box">
          <span>Monday</span>
          <h4>28</h4>
        </div>
      </div>
    </div>
  );
};
