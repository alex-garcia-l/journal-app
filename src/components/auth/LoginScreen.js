import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLogin, startLoginWithGoogle } from '../../actions/auth';

import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {

  const dispatch = useDispatch();
  const { loading } = useSelector(state => state.ui);

  const [valuesForm, handleInputChange] = useForm({
    email: '',
    password: ''
  });

  const { email, password } = valuesForm;

  const handleSubmit = (evt) => {
    evt.preventDefault();
    dispatch(startLogin(email, password));
  }

  const handleGoogleLogin = () => {
    dispatch(startLoginWithGoogle());
  }

  return (
    <div className="animate__animated animate__fadeIn animate__faster">
      <h3 className="auth__title mb-5">
        <i className="fa-solid fa-user-large fa-3x mb-5"></i>
        <span>Login</span>
      </h3>
      <form onSubmit={handleSubmit}>
        <input
          type="email"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="input"
          value={email}
          onChange={handleInputChange}
        />

        <input
          type="password"
          placeholder="Password"
          name="password"
          className="input"
          value={password}
          onChange={handleInputChange}
        />
        <button
          type="submit"
          className="btn btn-primary btn-block"
          disabled={loading}
        >
          Login
        </button>

        <div className="auth__social_networks">
          <p className="auth__paragraph">
            Login with social network
          </p>
          <div className="google-btn" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <div className="auth__create_acount">
          <Link
            to="/auth/register"
          >
            Create account
          </Link>
        </div>
      </form>
    </div>
  );
};
