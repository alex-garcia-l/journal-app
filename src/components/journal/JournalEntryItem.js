import { useDispatch } from 'react-redux';
import moment from 'moment';
import { setNoteActive } from '../../actions/note';

export const JournalEntryItem = ({ id, title, body, date, url }) => {

  const dispatch = useDispatch();
  const noteDate = moment(date);

  const note = {
    title,
    body,
    date,
    url
  }

  if (title.length > 30) {
    title = title.substring(0, 30) + '...';
  }

  if (body.length > 90) {
    body = body.substring(0, 90) + '...';
  }

  const handleClick = () => {    
    dispatch(setNoteActive(id, note));
  }

  return (
    <div
      className="journal__entry animate__animated animate__fadeIn"
      onClick={handleClick}
    >
      
      {
        url &&
        <div
          className="journal__entry-picture"
          style={{
            backgroundImage: `url(${url})`
          }}
        ></div>
      }

      <div className="journal__entry-container">
        <div className="journal__entry-body">
          <p className="journal__entry-title">
            { title }
          </p>

          <p className="journal__entry-content">
            { body }
          </p>
        </div>

        <div className="journal__entry-date-box">
          <span>{ noteDate.format('dddd') }</span>
          <h4>{ noteDate.format('Do') }</h4>
        </div>
      </div>
    </div>
  );
};
