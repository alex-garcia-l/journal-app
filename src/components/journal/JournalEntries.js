import { useSelector } from 'react-redux';
import { JournalEntryItem } from './JournalEntryItem';

export const JournalEntries = () => {

  const { notes } = useSelector( state => state.notes );

  return (
    <div className="journal__entries">
      {
        notes.map(note => (
          <JournalEntryItem
            key={note.id}
            { ...note }
          />
        ))
      }
    </div>
  );
};
