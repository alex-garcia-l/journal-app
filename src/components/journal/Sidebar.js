import { useDispatch, useSelector } from 'react-redux';

import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/note';
import { JournalEntries } from './JournalEntries';

export const Sidebar = () => {

  const dispatch = useDispatch();
  const { name } = useSelector(state => state.auth);

  const handleLogout = () => {
    dispatch(startLogout());
  }

  const handleNewNote = () => {
    dispatch(startNewNote());
  }

  return (
    <aside className="journal__sidebar">
      <nav className="journal__sidebar-navbar mt-5">
        <h3>
          <i className="far fa-moon"></i>
          <span className="ml-1">{name}</span>
        </h3>

        <button
          className="btn btn-outline-primary"
          onClick={handleLogout}
        >
          Logout
        </button>
      </nav>

      <div
        className="journal__new-entry"
        onClick={handleNewNote}
      >
        <i className="far fa-calendar-plus fa-4x"></i>
        <p className="mt-1">
          New entry
        </p>
      </div>

      <JournalEntries />
    </aside>
  );
};
