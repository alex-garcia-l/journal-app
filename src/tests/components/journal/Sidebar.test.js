/**
 * @jest-environment node
 */

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import { Sidebar } from '../../../components/journal/Sidebar';
import { startLogout } from '../../../actions/auth';
import { startNewNote } from '../../../actions/note';

import 'jsdom-global/register';

jest.mock('../../../actions/auth', () => ({
  startLogout: jest.fn(),
}));

jest.mock('../../../actions/note', () => ({
  startNewNote: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
  auth: {
    id: '123',
    name: 'Nombre de Usuario'
  },
  ui: {
    loading: false,
    msgError: null
  },
  notes: {
    active: null,
    notes: [],
  }
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <Sidebar />
  </Provider >
);

describe('Prueba en <Sidebar />', () => {

  beforeEach(() => {
    store = mockStore(initialState);
    jest.clearAllMocks();
  });

  test('debe mostrarse correctamente', () => {

    expect(wrapper).toMatchSnapshot();

  });

  test('debe llamar startLogout', () => {

    wrapper.find('button').simulate('click');

    expect(startLogout).toHaveBeenCalled();

  });

  test('debe llamar startNewNote', () => {

    wrapper.find('.journal__new-entry').simulate('click');

    expect(startNewNote).toHaveBeenCalled();

  });

});