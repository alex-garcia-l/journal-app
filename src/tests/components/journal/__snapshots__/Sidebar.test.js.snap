// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en <Sidebar /> debe mostrarse correctamente 1`] = `
<aside
  className="journal__sidebar"
>
  <nav
    className="journal__sidebar-navbar mt-5"
  >
    <h3>
      <i
        className="far fa-moon"
      />
      <span
        className="ml-1"
      >
        Nombre de Usuario
      </span>
    </h3>
    <button
      className="btn btn-outline-primary"
      onClick={[Function]}
    >
      Logout
    </button>
  </nav>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-4x"
    />
    <p
      className="mt-1"
    >
      New entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
